GW vasp

JosePizarro3 opened the pull request:
A few comments:

- Version >= 6.3.0 seems to be able to handle DFT+GW in one single mainfile, hence it will be good to generalize the VASPParser to handle these cases. But I have not examples and I am not sure whether it is enough to simply check the version to set the _child_archives.
- I re-defined / moved `parse_incarsinout` and `parse_kpoints` as methods of `VASPParser` to be callable by either `self.parse_method()` (DFT) or `self.parse_gw()` (GW).
- Fix of 'co**u**ntour_deformation' to 'contour_deformation' in some GW parsers.
- Defined `n_states_self_energy` as a complementary quantity to `n_empty_states_self_energy`: some codes cannot extract one or the other.

TODO:
- [X] Work on parsing from OUTCAR the GW eigenvalues. -> (not done) tbh, at this point this feature can be a future feature IF users ask for it. The OutcarTextParser is a bit broken as I could see, and cannot handle version>6.
- [x] Add tests for Si GW calculation.

JosePizarro3 committed:
Added gw.n_states_self_energy quantity for ABINIT and FHIaims, and fixed tests

JosePizarro3 committed:
Added parse_gw to VASP parser

Added child archives to VASP and extended for BeyondDFTWorkflowsParser

Added option for version > 6.3.0 to address single-step GW workflow calculations

Defined self.parse_kpoints()

JosePizarro3 committed:
Moved self.parse_incarsinout

Fixed parsing of k-mesh and q-mesh for GW

Added versioning for structure/nose

JosePizarro3 committed:
Deleted version>6.3.0 to handle GW calculations as it was buggy and not very generic

JosePizarro3 committed:
Deleted import BeyondDFTWorkflowsParser

JosePizarro3 commented:
@ladinesa could you please review this and the corresponding MR in nomad (1290)? It is pretty straightforward, and with this I can stop parsing and modifying the GW metainfo for a while.

@ndaelman-hu this might interest you, but I am not sure.

JosePizarro3 committed:
Added GW tests

Deleted nose structure _get_key_values, due to issues on versioning

JosePizarro3 commented:
Another remark:

- I deleted `nose` from the `get_structure` method, as it is being problematic. Maybe Joe can take a look in a future if he wants to implement MD calculations.

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -1014,13 +1014,15 @@ def get_structure(self, n_calc):
-                f'{structure}/nose[0]/v', array=True).get('v', None)
+        # COMMENTED OUT: there is a problem with the versioning: sometimes is nose
+        # others nose[0]
+        # if self._get_key_values(f'{structure}/nose/v', array=True):
```
maybe you could do an if nose/0/v

JosePizarro3 replied:
I tried, but it does not work. It seems to be `nose[0]/v` for >6.1.0

ladinesa replied:
maybe put a try to handle different versions.

JosePizarro3 replied:
It does not work because there is a `self.logger.error` when `_get_key_values` do not succeed, hence the error.

In any case, I still believe we can simply pass `nose=None` until Joe (if he will ever can) parses MD for VASP.

JosePizarro3 replied:
I added also a bit more on the comment to clarify this issue

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -1283,6 +1293,40 @@ def parse_method(self):
+        self.parse_incarsinout()
+
+        # Code-specific parameters
+        response_functions = self.parser._get_key_values(rf'/modeling[0]/parameters[0]/separator[@name="response functions"]/i')
```
this only works for vasprun.xml. You should implement get_response_functions in RunContentParser and also in OutcarContentParser

JosePizarro3 replied:
I tried to debug this, but it seems that the OUTCAR is never recognized as a mainfile.

For now, I will follow your advice and change to `self.parser.get_response_functions()` and pass when in the `OutcarContentParser`.

ladinesa replied:
OUTCAR is only recognized if there is no vasprun in the same directory. it is an alternative mainfile.

JosePizarro3 replied:
I see, I deleted the xml and now it works. Thanks!

ndaelman-hu replied:
Yes, but there are cases that it doesn't recognize the OUTCAR anyhow, even without vasprun.xml

ladinesa replied:
i suggest to investigate why. perhaps it is a version not captured by the current regex and needs adjustment.

JosePizarro3 replied:
Please, check out the new `get_response_functions` in the OutcarContentParser.

ladinesa commented:
Thanks for the changes. You can merge after addressing a couple of comments.

JosePizarro3 committed:
Added get_response_functions()

Deleted innecessary OUTCAR file from tests/data/vasp/Si_GW/

JosePizarro3 committed:
Added further comments on the nose versioning issue

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -680,6 +690,21 @@ def get_partial_dos(self, n_calc):
+                parameters_dict = {}
+                for param in parameters:
+                    if param[1] == 'T' or param[1] == 'F':
+                        param[1] = bool(param[1] == 'T')
```
is the conversion here necessary?

JosePizarro3 replied:
how can you pass 'T' as true and 'F' as false?

ladinesa replied:
simply param[1] == 'T'

JosePizarro3 replied:
ah, dumb me, you are totally right :smile:

JosePizarro3 replied:
done

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -680,6 +690,21 @@ def get_partial_dos(self, n_calc):
+        if self.parser.get('response_functions'):
+            parameters = self.parser.get('response_functions').get('input_parameters')
+            try:
+                parameters_dict = {}
```
put parameters_dict outside of if otherwise you get error in line 706

JosePizarro3 replied:
done

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -680,6 +690,21 @@ def get_partial_dos(self, n_calc):
         return dos, fields
 
+    def get_response_functions(self):
+        if self.parser.get('response_functions'):
```
if is not necessary if you set default to {}

JosePizarro3 replied:
done

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -680,6 +690,21 @@ def get_partial_dos(self, n_calc):
+            parameters = self.parser.get('response_functions').get('input_parameters')
+            try:
+                parameters_dict = {}
+                for param in parameters:
```
more compact `for param in self.parser.get('response_functions', dict()).get('input_parameters', dict())`

JosePizarro3 replied:
done

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -1118,74 +1151,75 @@ def get_partial_dos(self, n_calc):
         return dos, fields
 
+    def get_response_functions(self):
+        root = '/modeling[0]/parameters[0]/separator[@name="response functions"]/i'
```
put the handle for different versions here?

JosePizarro3 replied:
what do you mean different versions? The versioning problem is only in `node/` to my knowledge.

ladinesa replied:
o may bad i thougt i was looking at get_nose

JosePizarro3 replied:
no prob

JosePizarro3 committed:
Fix response_functions regex

Initialize parameters_dict out of try

Fix algo if whether extracted from xml (list) or OUTCAR (one parameters)

JosePizarro3 committed:
Added Alvin suggestions

JosePizarro3 committed:
Improved regex for response_functions.input_parameters thanks to GPT-4

JosePizarro3 commented:
Ok, finished: please @ladinesa resolve the conversations at will and I can merge.

There is a small detail: GPT-4 helped me improve the regex for these response_functions.input_parameters :smile:

ladinesa commented:
> Ok, finished: please @ladinesa resolve the conversations at will and I can merge.
> 
> There is a small detail: GPT-4 helped me improve the regex for these response_functions.input_parameters smile

ai is taking over our jobs...we're doomed.

JosePizarro3 committed:
Defined self.gw_algo to fix parse_gw

JosePizarro3 commented:
And another small improvement over `parse_gw`. 

From:
```
        # Type
        algo = self.parser.get_incar().get('ALGO')[0]
        nelmgw = response_functions.get('NELMGW', 1)
        if nelmgw == 1:
            sec_gw.type = 'G0W0'
        else:
            try:
                sec_gw.type = self._gw_algo_map[algo]
            except Exception:
                self.logger.warning('Error setting the GW sfc type.')
```

to (with a little my help re-defining algo):
```
        nelmgw = response_functions.get('NELMGW', 1)
        sec_gw.type = 'G0W0' if nelmgw == 1 else self._gw_algo_map.get(self.gw_algo, None)
        if sec_gw.type is None:
            self.logger.warning('Error setting the GW type.')
```

ladinesa commented:
> And another small improvement over `parse_gw`.
> 
> From:
> 
> ```
>         # Type
>         algo = self.parser.get_incar().get('ALGO')[0]
>         nelmgw = response_functions.get('NELMGW', 1)
>         if nelmgw == 1:
>             sec_gw.type = 'G0W0'
>         else:
>             try:
>                 sec_gw.type = self._gw_algo_map[algo]
>             except Exception:
>                 self.logger.warning('Error setting the GW sfc type.')
> ```
> 
> to (with a little my help re-defining algo):
> 
> ```
>         nelmgw = response_functions.get('NELMGW', 1)
>         sec_gw.type = 'G0W0' if nelmgw == 1 else self._gw_algo_map.get(self.gw_algo, None)
>         if sec_gw.type is None:
>             self.logger.warning('Error setting the GW type.')
> ```

The None default is not necessary hehe.

JosePizarro3 committed:
Alvin beated GPT-4
